File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
        std::queue<TreeNode*> nodes;
        nodes.push(root);
        while (nodes.size() > 0) {
            for (int i = 0; i < nodes.size(); i++) {
                TreeNode* current = nodes.front();
                minHeap.push(current->val);
                if (current->right) {
                    nodes.push(current->right);
                }
                if (current->left) {
                    nodes.push(current->left);
                }
                nodes.pop();
            }
        }
        for (int i = 0; i < k - 1; i++) {
            minHeap.pop();
        }
        return minHeap.top();
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        maxPath = [root.val]

        def depthFirstSearch(root):
            if not root:
                return 0

            leftMax = depthFirstSearch(root.left)
            rightMax = depthFirstSearch(root.right)

            leftMax = max(leftMax, 0)
            rightMax = max(rightMax, 0)

            maxPath[0] = max(maxPath[0], root.val + leftMax + rightMax)

            return root.val + max(leftMax, rightMax)

        depthFirstSearch(root)
        return maxPath[0]

